sparse-index.c

```diff
@@ -128,7 +133,7 @@ int convert_to_sparse(struct index_state *istate, int flags)
 {
 	int test_env;
 
-	if (istate->sparse_index || !istate->cache_nr ||
+	if (istate->sparse_index == COMPLETELY_SPARSE || !istate->cache_nr ||
```

Comment:
This isn't quite the right place for me to comment (the relevant line is `read-cache.c:1831`), but it still somewhat relates.

There seem to be situations where the index can be in a partially sparse (in this case meaning "partially collapsed") state when it's read in, but because it's always read in as `COMPLETELY_SPARSE` or `COMPLETELY_FULL`, checks like this don't end up collapsing it. For example, using the setup from `t1092`:

```
init_repos &&
git checkout -b test-branch update-folder1 &&
git reset --mixed HEAD~1 &&
echo a >folder1/a
```

At the end of this block, the index can be further collapsed (removing `folder1`, since it now contains no staged or unstaged changes), but the index will be read into the next command as `COMPLETELY_SPARSE`. 

TL;DR not as much for this PR as for sparse index in general, but is it valuable to make the initial read of the index determine whether it's `PARTIALLY_SPARSE` or `COMPLETELY_SPARSE` (or just always read as `PARTIALLY_SPARSE`)? Would that cause too much of a performance hit? If not, then my suggestion would be to add a note to the description of `PARTIALLY_SPARSE` that it's an internal-only state.

Reply:
You present a very interesting point: when we get these indexes that are not as collapsed as they could be, how do we get out of that state?

One way to do it would be to always assume `PARTIALLY_SPARSE` upon reading a new index. That currently leads to a performance drawback due to the recalculation of the cache-tree extension (perhaps we can avoid this?). I'll make a note to fix this later, as it is a problem that existed before the `PARTIALLY_SPARSE` mode.